### ISSUE TITLE
Slightly unclear about input for auto encoder

### ISSUE THREAD

**Issue by GlastonburyC** [erakra/DietNetworks#1]
Hi.

I'm very interested in the DAE application. I am trying to understand the format of your input data for this problem.
is the input ```(X) n samples (rows) by m SNPs (columns)```? (~2000 samples by 300k SNPs)

What format do the SNPs take? For each individual are they:
 ```-1 (missing), 0, 1, 2```?

How do you represent these as an input for the auto encoder - do you feed them in as they are (0, 1, 2), do you normalise them, or encode them in some other way?

Finally, is your input a single SNP across all samples (feature level row-wise), or is it the value of all SNPs per sample (sample-level, column-wise)?

Thanks!

**Comment by erakra** [comment]
Hi ! Thanks for asking ! Here are the answers to your questions

"is the input `(X) n samples (rows) by m SNPs (columns)` ?"
Yes !
"What format do the SNPs take?"
We use the SNP values (0, 1, 2). Let me just note again that we center and normalize each feature of the 300k features/SNPs)

Not sure to have understood your last question. So i am just trying to clarify what are the inputs of the whole model. X is the matrix defined as in your first question. The model has 2 inputs. X is the input to the classifier path (to predict the class of every sample, here the rows are the individuals), and X^T is the input to the auxiliary net that outputs the fat layer weight (here the rows are SNPs).

I hope it helps !

**Comment by GlastonburyC** [comment]
Hi @erakra. Thank you so much for taking the time to answer my questions! I appreciate it.

So all of my questions are just focusing on the auto encoder, not the classifier part.

So for the auto encoder - I wanted to know if your input is the SNPs per sample (350k different SNPs/features for an individual) or the SNP itself (i.e. the SNPs value across all individuals).

i.e. does the auto encoders first dense layer have 350k SNPs * number of hidden units = weights.

Thanks again!

**Comment by erakra** [comment]
Hi again,

The input to the auto-encoder and the classifier are the same (they actually share the encoder part of the auto-encoder). The input is X `n samples (rows) by m SNPs (columns)` (or a minibatch of it). So the input is the values of all the 350k SNPs per sample.

The first dense layer you are talking about is what we call the _fat_ layer in the paper. Yes it has 350k * n_units_layer1 as number of paramers. The goal of the suggested method is to generate such a weight matrix using much less free parameters.

I would advice checking the description of the model in the paper and the figure 1.

**Comment by GlastonburyC** [comment]
Hi - I enjoyed the paper and have read it a few times now.

Thanks for the clarification - I understand it's an auxiliary network learning the weights of the fat layer to minimise the parameter space - I was just interested in understanding the auxiliary network in isolation :)

Thanks.

**Comment by erakra** [comment]
Thanks again for your question. Glad you enjoyed the paper :+1: 

I was mainly adding more details to help future similar questions to be answered quickly (since it's a public issue ;) )

**Comment by GlastonburyC** [comment]
Ok no problem.

I'm still trying to clarify how the DAE fits into figure 1. From the text:

```training a denoising autoencoder (DAE) (Vincent et al., 2008) on the matrix *X* ```
Where X is defined in the paper as:  ```a matrix of data, with N samples and N_{d} features```

Can you try and elaborate here where the denoising autoencoder (SNP2vec) fits into the model architecture, and specifically how it helps reduce the parameter space? Because it seems to me that by training the autoencoder, you do have a fat layer but in the paper you report the SNP2vec model has only 10k parameters.

**Comment by erakra** [comment]
Ok, so concerning the SNP2vec , we train a DAE just to get an embedding that we use as a constant input to the auxiliary network (so the auxiliary net in this case does not use X^T as input). The number of parameters reported in the results concern the model trained for task we consider here (i.e the classifier and the auxiliary net that are trained end to end) which does not include the DAE weights in the case of SNP2vec.
One can think of it as follow : if you have a god-given embedding (here the snp2vec embedding) how can you deal with the large number of weights that will come along with the first layer of your classifier that may push it to overfit ? The fat layer of the DAE is not *directly* causing overfitting for the classification task.

So, in short, the SNP2vec DAE provides an embedding that you can inject as a fixed input to the auxiliary net to predict the weights of the fat layer of the classifier.

**Comment by GlastonburyC** [comment]
Thanks again!

Once you have the SNP2Vec encoder output - lets say each sample is now a 128-length vector SNP representation - Could you use that as the input to the classifier network (N samples by M DAE SNP vectors)?

Therefore the classifier network would have for its first dense layer, only 128 * n_hidden parameters?

I get a feeling you did this, compared it to the aux net predicted weight version, and that did better! :)
